### PR TITLE
Fix handling of unusual bundled plugin groupIds

### PR DESCRIPTION
**Pull request by svanoort**
Covers at least one case where there is a bundled plugin with groupId that is not org.jenkins-ci.plugins
Example: org.jenkins-ci.plugins.workflow plugins, such as workflow-step-api

For these plugins, the compat tester would generate pom files with groupId of org.jenkins-ci.plugins
With this bad groupId, it would fail to build, because artifact could not be found

Where available, we instead read the groupId from the plugin manifest.  

@reviewbybees - Resolves issues with github plugin and github pull request builder.

### CODE REVIEW

**Comment by amuniz** [comment]
:bee:

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 1]
```diff
@@ -150,8 +150,8 @@ public PluginCompatReport testPlugins()
             dataImporter = new DataImporter(config.getGaeBaseUrl(), config.getGaeSecurityToken());
         }
 
-
-        UpdateSite.Data data = config.getWar() == null ? extractUpdateCenterData() : scanWAR(config.getWar());
+        HashMap<String,String> pluginGroupIds = new HashMap<String, String>();  // Used to track real plugin groupIds from WARs
+        UpdateSite.Data data = config.getWar() == null ? extractUpdateCenterData() : scanWAR(config.getWar(), pluginGroupIds);
```
Comment (tfennelly) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: I always find these ternary operations painful to read :cry: . But I know this was here before you made your change.

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 2]
```diff
@@ -429,11 +429,21 @@ private TestExecutionResult testPluginAgainst(MavenCoordinates coreCoordinates,
...
+     * @return Update center data
+     * @throws IOException
+     */
+    private UpdateSite.Data scanWAR(File war, Map<String, String> pluginGroupIds) throws IOException {
         JSONObject top = new JSONObject();
         top.put("id", DEFAULT_SOURCE_ID);
         JSONObject plugins = new JSONObject();
         JarFile jf = new JarFile(war);
+        if (pluginGroupIds == null) {
+            pluginGroupIds = new HashMap<String, String>();
```
Comment (tfennelly) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: `Collections.emptyMap()` didn't work? Needs to be modifiable?

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 3]
```diff
@@ -626,17 +638,18 @@ private void addSplitPluginDependencies(String thisPlugin, MavenRunner.Config mc
...
-                        if (depBundledP != null) {
-                            System.out.println("Adding " + depPlugin + " since it was a dependency of " + plugin);
-                            toAdd.put(depPlugin, new VersionNumber(depBundledP.version));
-                        }
+                            Plugin depBundledP = otherPlugins.get(depPlugin);
+                            if (depBundledP != null) {
+                                System.out.println("Adding " + depPlugin + " since it was a dependency of " + plugin);
+                                toAdd.put(depPlugin, new VersionNumber(depBundledP.version));
+
+                            }
```
Comment (tfennelly) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: Did anything change here?

**Comment by tfennelly** [comment]
:bee: 

Not easily unit testable I assume?

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java (hunk 2 above)
Comment (svanoort) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: @tfennelly Correct, it's mutated within the function to build the list of items (if it's immutable there will be issues later on).  Not my favorite solution, but the alternative is to create a much more complex return type to this function just to encapsulate the extra map.

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java (hunk 3 above)
Comment (svanoort) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: Nope, looks like somehow a whitespace change slipped in *sigh.*  Pushing a commit to fix it.

**Comment by svanoort** [comment]
@reviewbybees done

**Comment by svanoort** [comment]
@tfennelly Yeah, unfortunately the whole library isn't very unit-test friendly (lots of interlinked functionality).  It's been manually tested though, with the WAR, and works (fixed what looked like a *real* issue too that was unrelated).

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java (hunk 2 above)
Comment (jglick) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: Why is this parameter not just `@Nonnull`? It was already initialized by the caller.

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 4]
```diff
@@ -511,10 +522,11 @@ private TestExecutionResult testPluginAgainst(MavenCoordinates coreCoordinates,
         }
     }
 
-    private void addSplitPluginDependencies(String thisPlugin, MavenRunner.Config mconfig, File pluginCheckoutDir, MavenPom pom, Map<String,Plugin> otherPlugins) throws PomExecutionException, IOException {
+    private void addSplitPluginDependencies(String thisPlugin, MavenRunner.Config mconfig, File pluginCheckoutDir, MavenPom pom, Map<String,Plugin> otherPlugins, Map<String, String> pluginGroupIds) throws PomExecutionException, IOException {
         File tmp = File.createTempFile("dependencies", ".log");
         VersionNumber coreDep = null;
         Map<String,VersionNumber> pluginDeps = new HashMap<String,VersionNumber>();
+        Map<String, String> pluginGroups = new HashMap<String, String>();
```
Comment (jglick) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: What is this for? Appears unused.

**Comment by jglick** [comment]
You forgot to fix the other place where `org.jenkins-ci.plugins` was hard-coded.

**Comment by svanoort** [comment]
I'm guessing this will want to be opened in a separate PR if needed, but with regard to the last comment:

@jglick You're talking about this:  https://github.com/jenkinsci/plugin-compat-tester/blob/master/plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java#L554  ?  Do we would want to glob off of this to anything starting with org.jenkins-ci.plugins i.e. org.jenkins-ci.plugins*)?  (Not sure how else to ID actual plugins). 

This part I'd held off changing in the initial pass because I wasn't sure if it might open up to other edge cases, but if the approach above sounds good, I've got something ready to out in a new PR (tests running right now).

**Comment by jglick** [comment]
> Do we would want to glob off of this to anything starting with org.jenkins-ci.plugins

No, I was thinking rather that you would leave the existing branches as they are (as a fallback, especially for when `-war` is not used), but then add something like

```java
} else if (groupId.equals(groupIds.get(artifactId))) {
    pluginDeps.put(artifactId, version);
```
